src/content/en/updates/2017/02/media-session.md

```diff
@@ -0,0 +1,398 @@
+- `Else` fetch artwork from network
+    - `If` fetch is successful, add network artwork to the cache and serve it
+    - `Else` serve fallback artwork from the cache
+
+That way, media notifications would always have a nice artwork icon even when
+browser can't fetch it. Here's how you could implement this for instance:
+
+    const FALLBACK_ARTWORK_URL = 'fallbackArtwork.png';
+    
+    self.addEventListener('install', event => {
+      self.skipWaiting();
+      event.waitUntil(initArtworksCache());
+    });
+    
+    function initArtworksCache() {
+      caches.open('artworks-cache-v1')
+      .then(cache => cache.add(FALLBACK_ARTWORK_URL));
+    }
+    
+    self.addEventListener('fetch', event => {
+      if (/artwork-[0-9]+\.png$/.test(event.request.url)) {
+        event.respondWith(handleFetchArtwork(event.request));
+      }
+    });
+    
+    function handleFetchArtwork(request) {
+      // Return cache request if it's in the cache already, otherwise fetch
+      // network artwork.
+      return getCacheArtwork(request)
+      .then(cacheResponse => cacheResponse || getNetworkArtwork(request));
+    }
+    
+    function getCacheArtwork(request) {
+      return caches.open('artworks-cache-v1')
+      .then(cache => cache.match(request));
+    }
+    
+    function getNetworkArtwork(request) {
+      // Fetch network artwork.
+      return fetch(request)
+      .then(networkResponse => {
+        if (networkResponse.status !== 200) {
+          return Promise.reject('Network artwork response is not valid');
+        }
+        // Add artwork to the cache for later use and return network response.
+        addArtworkToCache(request, networkResponse.clone())
+        return networkResponse;
+      })
+      .catch(error => {
+        // Return cached fallback artwork.
+        return getCacheArtwork(new Request(FALLBACK_ARTWORK_URL))
+      });
+    }
+    
+    function addArtworkToCache(request, response) {
+      return caches.open('artworks-cache-v1')
+      .then(cache => cache.put(request, response));
+    }
+
+Caution: If you want your service worker to be able to intercept artwork
+network requests on [the very first page load], you may want to call
+`clients.claim()` within your service worker once it's activated.
+
+### Let user control cache
+
+As user will consume content from your web app, media files and artworks may
+take a lot of space on user's device. It is **your responsibility to show how
+much cache is used and give user the ability to clear it**. Thankfully for us,
+doing so is pretty easy with the [Cache API].
+
+    // Here's how I'd compute how much cache is used by artworks... 
+    caches.open('artworks-cache-v1')
+    .then(cache => cache.matchAll())
+    .then(responses => Promise.all(responses.map(r => r.blob())))
+    .then(blobs => blobs.map(blob => blob.size).reduce((acc, cur) => acc + cur, 0))
+    .then(cacheSize => { console.log('Artworks cache is ' + cacheSize + ' bytes.'); })
```

Comment:
This is going to bring all the artworks into memory. Alternatively you could add up the `content-length` headers, and make estimates where this header is missing.